services/gundeck/test/unit/MockGundeck.hs

```diff
@@ -0,0 +1,716 @@
+  maybe (pure []) (pure . Map.elems) $ Map.lookup uid getaddr
+
+mockBulkSend
+  :: (HasCallStack, m ~ MockGundeck)
+  => URI -> BulkPushRequest
+  -> m (URI, Either SomeException BulkPushResponse)
+mockBulkSend uri notifs = do
+  getstatus <- mkWSStatus
+  let flat :: [(Notification, PushTarget)]
+      flat = case notifs of
+        (BulkPushRequest ntifs) ->
+          mconcat $ (\(ntif, trgts) -> (ntif,) <$> trgts) <$> ntifs
+
+  forM_ flat $ \(ntif, ptgt) -> do
+    when (getstatus ptgt == PushStatusOk) $ do
+      modify $ msWSQueue %~ \queue ->
+        deliver queue (ptUserId ptgt, clientIdFromConnId $ ptConnId ptgt) (ntfPayload ntif)
+
+  pure . (uri,) . Right $ BulkPushResponse
+    [ (ntfId ntif, trgt, getstatus trgt) | (ntif, trgt) <- flat ]
+
+mockOldSimpleWebPush
+  :: (HasCallStack, m ~ MockGundeck)
+  => Notification
+  -> List1 NotificationTarget
+  -> UserId
+  -> Maybe ConnId
+  -> Set ConnId
+  -> m [Presence]
+mockOldSimpleWebPush _ _ _ _ (Set.null -> False) =
+  error "connection whitelists are not implemented in this test."
+  -- (see 'genPush' above)
+
+mockOldSimpleWebPush notif tgts _senderid mconnid _ = do
+  getrecp   <- mkGetRecipient
+  getstatus <- mkWSStatus
+
+  let clients :: [(UserId, ClientId)]
+      clients
+        = -- reformat
+          fmap (\(PushTarget uid connid) -> (uid, clientIdFromConnId connid))
+          -- drop all broken web sockets
+        . filter ((== PushStatusOk) . getstatus)
+          -- do not push to sending device
+        . filter ((/= mconnid) . Just . ptConnId)
+          -- reformat
+        . mconcat . fmap (\tgt ->
+                            PushTarget (tgt ^. targetUser) . fakeConnId
+                              <$> (tgt ^. targetClients))
+        . fmap emptyMeansFullHack
+        $ toList tgts
+
+      emptyMeansFullHack :: NotificationTarget -> NotificationTarget
+      emptyMeansFullHack tgt = tgt & targetClients %~ \case
+        []   -> getrecp (tgt ^. targetUser) ^. recipientClients
+        same@(_:_) -> same
+
+  forM_ clients $ \(userid, clientid) -> do
+    modify $ msWSQueue %~ \queue ->
+      deliver queue (userid, clientid) (ntfPayload notif)
+
+  pure $ uncurry fakePresence <$> clients
+
+
+----------------------------------------------------------------------
+-- helpers
+
+-- | (it may be possible to drop this type in favor of more sophisticated use of quickcheck's
+-- counterexample.)
+newtype Pretty a = Pretty a
+  deriving (Eq, Ord)
+
+instance Aeson.ToJSON a => Show (Pretty a) where
+  show (Pretty a) = cs $ Aeson.encodePretty a
+
+shrinkPretty :: (a -> [a]) -> Pretty a -> [Pretty a]
+shrinkPretty shrnk (Pretty xs) = Pretty <$> shrnk xs
+
+
+deliver :: NotifQueue -> (UserId, ClientId) -> Payload -> NotifQueue
+deliver queue qkey qval = Map.alter (Just . tweak) qkey queue
+  where
+    tweak Nothing      = Set.singleton (payloadToInt qval)
+    tweak (Just qvals) = Set.insert    (payloadToInt qval) qvals
+
+payloadToInt :: Payload -> Int
```

Comment:
Hm, why `toRational (x * 100)` and not just taking the number in the payload?

Also, do we test invalid payloads somewhere? If not, it might be better to just redefine `newtype Payload = Payload Int`, which would have the benefit of immediately signalling that in mock Gundeck we use only a subset of possible payloads.

Reply:
good idea, but:

```haskell
-- | We really don't care about the actual payloads anywhere in these tests, just that the right
-- ones arrive over the right connections.  So 'genPayload' is not very exhaustive, but only
-- generates small objects with one field containing a numeric value.  It would be nice to represent
-- this in the 'Payload' type, but the 'List1 Aeson.Object' structure is used in the production
-- code, so in the end it is more awkward than nice.
type Payload = List1 Aeson.Object

[...]

-- | Return the rounded number contained in the payload.  This is enough to detect message mixups,
-- but less noisy in the printed counter-examples.
payloadToInt :: Payload -> Int
```

ok?

Reply:
> in the end it is more awkward than nice

Alright.

> rounded number

I still don't understand what's the rationale behind multiplying an always-integer number by 100, but I guess it doesn't hurt either.

Reply:
> I still don't understand what's the rationale behind multiplying an always-integer number by 100, but I guess it doesn't hurt either.

i think i noticed numbers that only differed in their tenths and hundredths, and i wanted to keep more information after rounding.

Reply:
I think I solved the mystery – `Arbitrary Aeson.Value` could be doing that. We don't have it anymore, but if fractional payloads ever creep into the tests again, `toBoundedInteger` will catch them so I feel pretty safe.